src/pudl/extract/extractor.py

```diff
@@ -0,0 +1,379 @@
+        if len(partition_names) != 1:
+            raise AssertionError(
+                f"Expecting exactly one attribute to define this partition (found: {partition})"
+            )
+
+        partition_name = partition_names[0]
+        partition_selection = partition[partition_name]
+        if isinstance(partition_selection, list | tuple):
+            raise AssertionError(
+                f"Expecting exactly one non-container value for this partition attribute (found: {partition})"
+            )
+        return str(partition_selection)
+
+    def get_all_pages(self) -> list[str]:
+        """Returns list of all known pages."""
+        return sorted(self._column_map.keys())
+
+    def get_all_columns(self, page) -> list[str]:
+        """Returns list of all pudl columns for a given page across all partitions."""
+        return sorted(self._column_map[page].T.columns)
+
+
+class GenericExtractor:
+    """Generic extractor base class."""
+
+    METADATA: GenericMetadata = None
+    """Instance of metadata object to use with this extractor."""
+
+    BLACKLISTED_PAGES = []
+    """List of supported pages that should not be extracted."""
+
+    def __init__(self, ds):
+        """Create new extractor object and load metadata.
+
+        Args:
+            ds (datastore.Datastore): An initialized datastore, or subclass
+        """
+        if not self.METADATA:
+            raise NotImplementedError("self.METADATA must be set.")
+        self._metadata = self.METADATA
+        self._dataset_name = self._metadata.get_dataset_name()
+        self.ds = ds
+        self.cols_added: list[str] = []
+
+    @abstractmethod
+    def source_filename(self, page: str, **partition: PartitionSelection) -> str:
+        """Produce the source file name as it will appear in the archive.
+
+        Args:
+            page: pudl name for the dataset contents, eg "boiler_generator_assn" or
+                "coal_stocks"
+            partition: partition to load. Examples:
+                {'year': 2009}
+                {'year_month': '2020-08'}
+
+        Returns:
+            string name of the source file
+        """
+        ...
+
+    @abstractmethod
+    def load_source(self, page: str, **partition: PartitionSelection) -> pd.DataFrame:
+        """Produce the source data for the given page and partition(s).
+
+        Args:
+            page: pudl name for the dataset contents, eg
+                "boiler_generator_assn" or "coal_stocks"
+            partition: partition to load. Examples:
+                {'year': 2009}
+                {'year_month': '2020-08'}
+
+        Returns:
+            pd.DataFrame instance with the source data
+        """
+        ...
+
+    def process_raw(
+        self, df: pd.DataFrame, page: str, **partition: PartitionSelection
+    ) -> pd.DataFrame:
+        """Takes any special steps for processing raw data and renaming columns."""
+        return df
+
+    @staticmethod
```

Comment:
`@staticmethod` annotations were pulled from the original, less abstract implementation. I'm not clear on the original motivation. For `process_renamed` at least the method (including subclass implementations) is only ever called at a point where an extractor object would have been instantiated.

Reply:
I'm not sure either, it doesn't make intuitive sense to me - maybe @zschira you have thoughts?